vpr/src/noc/north_last_routing.cpp

```diff
@@ -96,4 +96,29 @@ TurnModelRouting::Direction NorthLastRouting::select_next_direction(const std::v
     }
 
     return selected_direction;
-}
\ No newline at end of file
+}
+
+bool NorthLastRouting::is_turn_legal(const std::array<std::reference_wrapper<const NocRouter>, 3>& noc_routers) const {
+    const int x1 = noc_routers[0].get().get_router_grid_position_x();
+    const int y1 = noc_routers[0].get().get_router_grid_position_y();
+
+    const int x2 = noc_routers[1].get().get_router_grid_position_x();
+    const int y2 = noc_routers[1].get().get_router_grid_position_y();
+
+    const int x3 = noc_routers[2].get().get_router_grid_position_x();
+    const int y3 = noc_routers[2].get().get_router_grid_position_y();
+
+    VTR_ASSERT(x2 == x1 || y2 == y1);
+    VTR_ASSERT(x3 == x2 || y3 == y2);
+
+    // going back to the first router is not allowed
+    if (x1 == x3 && y1 == y3) {
+        return false;
+    }
+
+    if (y2 > y1 && x2 != x3) {
```

Comment:
Comment: Check that we're following the North last algorithm (can't go up and then another direction)